File: labs/lab1.2/csharp/MyString.cs
namespace Program;

// Значення рядка (цифрові символи),
// конструктори, деструктор, метод
// обчислення довжини, метод обернення
// рядка, метод отримання значення рядка)
class MyString {
    private string _value;
    public string Value {
        get => this._value;
        set {
            if(!MyString.IsNumeric(value)) throw new Exception("String must be numeric");
            this._value = value;
        }
    }

    public MyString() {
        this._value = "";
    }
    public MyString(string text) {
        this.Value = text;
    }
    public MyString(char[] text) {
        this.Value = new String(text);
        
    }
    ~MyString() {
        Console.WriteLine($"String {this._value} is destroyed");
    }

    public int Lenght() => this._value.Length;

    public void Inverse() {
        char[] chars = this._value.ToCharArray();
        Array.Reverse(chars);
        this._value = new String(chars);
    }

    public IEnumerator<char> GetEnumerator() {
        foreach(char c in this._value) yield return c;
    }

    public override string ToString() => this._value;

    public static bool IsNumeric(string text) {
        foreach(char i in text) {
            if(!Char.IsDigit(i)) return false;
        }
        return true;
    }

    public static MyString operator +(MyString a, MyString b) => new MyString(a._value+b._value);
    public char this[int key] {
        get => this._value[key];
    }
}

File: labs/lab1.2/typescript/MyString.ts
// Значення рядка (цифрові символи),
// конструктори, деструктор, метод
// обчислення довжини, метод обернення
// рядка, метод отримання значення рядка)

export class MyString {
    private value: string;
    public get Value():string {return this.value}
    public set Value(text:string) {
        if(!MyString.IsNumeric(text)) throw new Error("String must be numeric");
        this.value = text
    }

    constructor()
    constructor(text:string)
    constructor(text:MyString)
    constructor(text?:string|MyString) {
        if (typeof text == "string") {
            this.Value = text
        } else if (text instanceof MyString) {
            this.value = text.value
        } else {
            this.value = ""
        }
    }

    public Length():number {return this.value.length}

    public Inverse():void {this.value = this.value.split("").reverse().join("")}

    public Concat(other:MyString):MyString {return new MyString(this.value + other.value)}

    public CharAt(index:number):string {return this.value.charAt(index)}

    public static IsNumeric(text: string): boolean {
        for (const c of text) {
            if (isNaN(Number(c))) return false
        }
        return true;
    }
}
